### ISSUE TITLE
MyPaint slow/lagging once symmetry tool is enabled 

### ISSUE THREAD

**Issue by hzulla** [mypaint/mypaint#883]
#### Description of the problem

Using MyPaint on a relatively slow RPi3 computer with a Wacom tablet works fine. Unless the user activates the symmetry tool. Once adding a symmetry line to the drawing, the drawing on the screen significantly lags behind the tablet pen movements.

#### Basic system details

MyPaint 1.2.0 on Raspbian Stretch (November 2017) using Raspberry Pi Desktop

#### Steps to reproduce

1. Connect Wacom tablet to RPi3
2. Start MyPaint
3. Draw a little, notice the screen drawing not lagging (at least, not much)
4. Add symmetry line
5. Draw again, notice the screen drawing lagging a lot

**Comment by achadwick** [comment]
Symmetry mode necessitates more work from the brush engine, that's just a fact of life.

Q: does the laggy behaviour go away after the symmetry mode is turned off?

I would normally suggest using a newer MyPaint, but I'm not sure that would really help you with this particular machine.

Whoever tries to deal with this should probably have a look at [DEBUGGING.md#Profiling](https://github.com/mypaint/mypaint/blob/bae0db944dc1dd2afd7a5026d36cd395bd4f372e/DEBUGGING.md) to get started.

**Comment by hzulla** [comment]
Yes, the lag goes away when the symmetry tool is inactive again.

**Comment by briend** [comment]
For anyone looking at this, try turning on Debug-->Visualize Rendering.   With symmetry mode active the bounding box for redrawing the dirty tiles becomes  bigger as you move away from the center line.  If you turn on horizontal and vertical the problem becomes exponential as you very quickly begin to redraw the entire screen.  Ouch!

I don't know how to begin fixing this. . .  maybe make the brush engine return separate bounding boxes for each dab draw instead of one giant bounding box.

**Comment by jplloyd** [comment]
> For anyone looking at this, try turning on Debug-->Visualize Rendering. With symmetry mode active the bounding box for redrawing the dirty tiles becomes bigger as you move away from the center line. 

Yep, the problem is indeed in the redrawing in MyPaint. libmypaint does not even break a sweat when rendering e.g. rotational symmetry with 50 rotational lines (unless the brush itself is heavy).

>If you turn on horizontal and vertical the problem becomes exponential as you very quickly begin to redraw the entire screen. Ouch!

Not quite exponential, but a very nice visual example of quadratic complexity ;-)

> I don't know how to begin fixing this. . . maybe make the brush engine return separate bounding boxes for each dab draw instead of one giant bounding box.

Funnily enough, @jonnor suggested this in a comment in libmypaint/brushlib over 7 years ago, but I don't think it was ever implemented (no need at the time, I guess). At any rate, I implemented a naive version of multi-rectangle bounding boxes today, and while there is still _some_ lag on fast strokes, it's a huge improvement for the new symmetry modes overall!

**Comment by jplloyd** [comment]
The quadratic behavior is fixed with commits 0dd0de899a919a96deb9ecb686527f36f439ad82 and mypaint/libmypaint@ec0447ea54552a350119013d624cb0b45582155c

The overall performance is still slower than it needs to be, due to the previously mentioned bottlenecks, but that is not related directly to the symmetry modes (the slowness of the mode is just a symptom of the way redrawing is handled).

**Comment by jplloyd** [comment]
Actually, I just remembered that there still _are_ performance issues related more closely to the symmetry mode. The fix I referenced above don't fix e.g. line mode being very slow when symmetry is enabled. Presumably the inking mode is just as bad.
